pennylane/pytrees.py

```diff
@@ -52,7 +104,104 @@ def register_pytree(pytree_type: type, flatten_fn: FlattenFn, unflatten_fn: Unfl
     Side Effects:
         ``pytree`` type becomes registered with available backends.
 
+    .. seealso:: :func:`~.flatten`, :func:`~.unflatten`.
+
     """
 
+    _register_pytree_with_pennylane(pytree_type, flatten_fn, unflatten_fn)
+
     if has_jax:
         _register_pytree_with_jax(pytree_type, flatten_fn, unflatten_fn)
+
+
+@dataclass(repr=False, frozen=True)
+class PyTreeStructure:
+    """A pytree data structure, holding the type, metadata, and child pytree structures.
+
+    >>> op = qml.adjoint(qml.RX(0.1, 0))
+    >>> data, structure = qml.pytrees.flatten(op)
+    >>> structure
+    PyTree(AdjointOperation, (), [PyTree(RX, (<Wires = [0]>, ()), [Leaf])])
+
+    A leaf is defined as just a ``PyTreeStructure`` with ``type=None``.
+    """
+
+    type: Optional[type]
+    """The type corresponding to the node. If ``None``, then the structure is a leaf."""
+
+    metadata: Metadata
+    """Any metadata needed to reproduce the original object."""
+
+    children: list["PyTreeStructure"]
+    """The children of the pytree node.  Can be either other structures or terminal leaves."""
+
+    def __repr__(self):
+        if self.type is None:
+            return "Leaf"
+        return f"PyTree({self.type.__name__}, {self.metadata}, {self.children})"
```

Comment:
With the defaults added above, we can have a more informative repr that should also be eval() compatible:

```suggestion
    def __repr__(self):
        if self.type is None:
            return f"PytreeStructure()"
            
        return f"PyTreeStructure({self.type.__name__}, {repr(self.metadata)}, {repr(self.children)})"
```

Reply:
Here we have to choose between something more readable and something more eval-compatible.

Let's go with the eval-compatible option for the repr, but then also add the more readable option as str.